Efficient sequence alignment path data structure

mataton opened the pull request:
This PR will implement the sequence alignment path data structure as outlined in issue #1974. 

Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [x] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

mataton commented:
@qiyunzhu @mortonjt This PR is more than 90% completed. A preliminary review would be appreciated.

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+class AlignPath(SkbioObject):
+    def __init__(self, lengths, states, starts):
+        """Create an alignment path from segment lengths and states.
```
```suggestion
        r"""Create an alignment path from segment lengths and states.
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        """
+        # Number of sequences needs to be explicitly provided, because the packed bits
+        # does not contain this information. (It is merely in multiples of 8.)
+        self.lengths = np.asarray(lengths, dtype=int)
```
```suggestion
        self.lengths = np.asarray(lengths, dtype=dtype)
```
I wonder if we should have a `dtype` flag to be passed in, and have `np.uint32` be the default. 
But again, I'm not sure if this data structure is sufficient for genome-scale alignments (it can't recombination right?) Otherwise, these alignments could get *huge* (i.e. trillion bp genomes seen in some plants)

qiyunzhu replied:
This might be a good point. It may not be a concern in Linux or MacOS, but in Windows (even 64-bit), `int` defaults to `np.int32`, the maximum of which is 2,147,483,647 (i.e., 2 Gbp), which is shorter than many genomes. Perhaps we can default it to `np.uint64`. I am not sure if we need to create a dtype flag. This may confuse non-technical users.

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+        # TODO: Needs to think about whether reverse complemented (Boolean array of
+        # (n_seqs,)) should be included as a parameter. It is only relevant for
+        # nucleotide sequences.
```
I think we can scrap this requirement -- users should in theory reverse complement their own sequences, right?

qiyunzhu replied:
They usually should. Although multiple programs like BLAST and Bowtie2 have the option to search both forward and reverse strands.

mataton replied:
I can scrap it for now. If we want to implement in the future we always can.

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        # of the alignment in each sequence (shape: (n_seqs,)). This is important
+        # especially for local alignments (e.g., search for a short read in a genome
+        # sequence). If it is provided, `n_seqs` is no longer necessary.
+        # self.starts = starts
```
Has this been addressed?

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        """Return string representation of this AlignPath."""
+        # Not sure if this makes sense for this class, but it is needed for all
+        # SkbioObjects.
+        return self.__repr__()
```
It may not be a bad idea to return the full CIGAR string for a pairwise alignment. Not sure what the analogy is for an MSA

mortonjt replied:
Actually, the `to_cigar` method would round this out.

qiyunzhu replied:
I don't know if there is a CIGAR equivalent for multiple sequence alignments. I guess our `AlignPath` is the closest data structure.

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+    @classonlymethod
+    def from_tabular(cls, msa):
+        """Create an alignment path from a `TabularMSA` object."""
```
Is it worthwhile to have a `to_tabular` version method? It maybe ideal for easier manipulation / visualization

mortonjt replied:
```suggestion
        r"""Create an alignment path from a `TabularMSA` object."""
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+        Better to have ability to index an AlignPath for particular sequences,
+        then compress these sequences using something like _fix_arrays"""
+        pass
```
Should we delete this, since it isn't implemented? We can raise a `NotImplementedError` otherwise

mataton commented on skbio/alignment/_util.py:
```diff
@@ -0,0 +1,51 @@
+import numpy as np
```
@mortonjt @qiyunzhu This file is incomplete and untested, feel free to ignore for now.

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+from skbio.util._decorator import overrides, classonlymethod
+
+
+class AlignPath(SkbioObject):
```
Is it worthwhile to flesh out a practical use-case of this object? Pseudocode in the issue tracker may be useful to help us understand how this code is going to be used (i.e. pseudocount for progressive alignment)

mataton replied:
@qiyunzhu You may have a clearer or more well defined use case in mind than I do?

qiyunzhu replied:
I think we should do it but it is outside the scope of the current PR. Specifically, we can:
1. Generate a pairwise alignment using external Python libraries like BioPython or Biotite and convert the result into `AlignPath`.
2. Generate a multiple alignment using external programs like ClustalO or MAFFT and read the result (multi-fasta format) into `AlignPath`.
3. Perform pairwise or multiple alignment using scikit-bio iteself (to be re-modeled) and return an `AlignPath` object.
4. Manually create a plausible alignment and calculate its alignment score.

There may be variants of each case. For example, for 3, we can return all optimal alignments instead of just one arbitrary alignment (pseudocode):

```python
for aln in pairwise_align(seq1, seq2, return_all=True):
    scores.append(aln.align_score())
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+    @classonlymethod
+    def from_bits(cls, bits):
+        """Create an alignment path from a bit array (0 - char, 1 - gap)."""
```
```suggestion
        r"""Create an alignment path from a bit array (0 - char, 1 - gap)."""
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        return cls.from_bits(np.isin(byte_arr, gap_chars))
+
+    def to_indices(self, gap=-1):
+        """Generate an array of indices of characters in the original sequences.
```
```suggestion
        r"""Generate an array of indices of characters in the original sequences.
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+    @classonlymethod
+    def from_indices(cls, indices, gap=-1):
+        """Create an alignment path from character indices in the original sequences.
```
```suggestion
        r"""Create an alignment path from character indices in the original sequences.
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+            return cls.from_bits(indices == gap)
+
+    def to_coordinates(self):
+        """Generate an array of segment coordinates in the original sequences."""
```
```suggestion
        r"""Generate an array of segment coordinates in the original sequences."""
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        return lens.cumsum(axis=1)
+
+    @classonlymethod
+    def from_coordinates(cls, coords):
```
```suggestion
    def from_coordinates(cls, coords):
        r"""Generate the an alignment path from an array of segment coordinates."""
```
Don't forget the docstring, this won't be documented otherwise

mortonjt replied:
Also need to document the parameteres

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+    @overrides(AlignPath)
+    @classonlymethod
+    def from_bits(cls, bits):
+        """Create an alignment path from a bit array."""
```
```suggestion
        r"""Create an alignment path from a bit array."""
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+    def from_bits(cls, bits):
+        """Create an alignment path from a bit array."""
+        # This should be faster than the generic solution I guess.
+        # TODO: Pending benchmark and optimization.
```
Should we perform a benchmark before merging?

mataton replied:
Yes, I can talk with @qiyunzhu in person and flesh this out.

qiyunzhu replied:
I benchmarked the conversion part on real-world data and it is pretty fast. Matt added the validation.

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+    def to_cigar(self, seqs=None):
+        """Get a CIGAR string representing the pairwise alignment path.
+
+        seqs: If provided, will distinguish match (=) and mismatch (X). Otherwise,
```
```suggestion
    Parameters
    ----------
        seqs: If provided, will distinguish match (=) and mismatch (X). Otherwise,
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        return np.stack([self.states & 1, self.states >> 1])
+
+    def to_cigar(self, seqs=None):
+        """Get a CIGAR string representing the pairwise alignment path.
```
```suggestion
        r"""Get a CIGAR string representing the pairwise alignment path.
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+    @classonlymethod
+    def from_cigar(cls, cigar):
+        """Create a pairwise alignment path from a CIGAR string."""
```
```suggestion
        r"""Create a pairwise alignment path from a CIGAR string."""
```

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+        return cls(np.asarray(lengths), np.asarray(gaps), [0, 0])
+
+    def _run_length_encode(self, input_string):
+        """Perform run length encoding on a string."""
```
```suggestion
        r"""Perform run length encoding on a string."""
```
Last one, make sure all of the docstrings are formatted appropriately

mortonjt commented on skbio/alignment/_tabular_msa.py:
```diff
@@ -2477,3 +2477,16 @@ def _is_sequence_axis(self, axis):
         return self.shape.position
+
+
+class TabularMSA2(TabularMSA):
```
Can you clarify where this is coming from? Do we need this? Or can we just include this in `TabularMSA`?

mataton replied:
This will be included in `TabularMSA`, will address this in future commit.

mortonjt commented on skbio/alignment/_util.py:
```diff
@@ -0,0 +1,51 @@
+import numpy as np
+
+
+def align_score(path, seqs, match=2, mismatch=-3, gap_open=-5, gap_extend=-2):
```
There are already quite a few alignment methods in skbio, so we need to differientiate. Maybe `pairwise_align_score`?

mortonjt commented on skbio/alignment/_util.py:
```diff
@@ -0,0 +1,51 @@
+    )
+
+
+def align_scores(path, seqs, match, mismatch, gap_open, gap_extend):
```
Same, maybe `multiple_align_score`?

mortonjt commented on skbio/alignment/_util.py:
```diff
@@ -0,0 +1,51 @@
+    pass
+
+
+def align_sp_scores(path, seqs, match, mismatch, gap_open, gap_extend):
```
Maybe `multiple_align_sp_score`

mortonjt replied:
What do you think of having a `multiple_dm_score` that instead passes in distance matrix (which could be sum-of-pairs, or any distance metric). This could look like

`def align_sp_scores(path, seqs, match_dm, mismatch, gap_open, gap_extend):`

Smith-waterman can work with just gaps + a matrix distance matrix (either blosum, or an index-specific distance matrix).

mortonjt commented on skbio/alignment/_util.py:
```diff
@@ -0,0 +1,51 @@
+    pass
+
+
+def separate_rle(msa):
```
doc?

qiyunzhu replied:
This `separate_rle` is a completely different data structure. Before we sufficiently explore it, I would suggest that we don't include it in this PR.

mortonjt commented on skbio/alignment/__init__.py:
```diff
@@ -218,11 +218,14 @@
     local_pairwise_align_ssw,
 )
 from skbio.alignment._ssw_wrapper import StripedSmithWaterman, AlignmentStructure
+from skbio.alignment._path import AlignPath, PairAlignPath
```
Make sure that the `toctree` above is documented. Otherwise these objects won't show up in the docs

mortonjt commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,308 @@
+
+class PairAlignPath(AlignPath):
+    @overrides(AlignPath)
+    @classonlymethod
```
It looks like some of the unittests still throws an error for python=3.8
```
AttributeError: 'classonlymethod' object has no attribute '__name__'
```
Not sure why this is being thrown, but we'll want to investigate this. Does this need the `@overrides(AlignPath)` decorator?

mataton replied:
Removing the `overrides` decorator solved this issue without affecting functionality.

mortonjt reviewed (changes requested):
Addressed some comments. Its a relatively incomplete review, so we'll want at least one more pass.

Also, don't forget to add this to the `index.rst` -- this won't show up in the documentation otherwise.

mataton commented:
@qiyunzhu One question I have regarding `align_scores` or calculating a matrix of pairwise alignment scores from an MSA: If at a single position in the pairwise alignment there are gaps in each sequence, how is this scored? From the example in the original [issue](https://github.com/scikit-bio/scikit-bio/issues/1974):

````python
seq2 = 'CAG--GTAAG-CATACCTCA'
seq3 = 'CGGTCGTCAC-TGTACACTA'
````

At the position where both sequences contain a gap, how is this interpreted or scored?

mataton commented on skbio/alignment/_util.py:
```diff
@@ -0,0 +1,70 @@
+    # based on the characters within the non-gap segments
+    # This will work for a strictly pairwise alignment, where it is certain that
+    # gaps will not occur in the same position in both sequences.
+    seq1, seq2 = (x._bytes for x in seqs)
```
@qiyunzhu This code is much simpler than the original, which has been commented out below. I think the original intended to account for situations where there are gaps in the same position in the two sequences. My understanding is that in a purely pairwise alignment, this will never occur. Is that correct? I feel this function should just be about a straightforward pairwise alignment score. Handling other cases could be put into the other functions in this file. What do you think?

qiyunzhu commented on skbio/alignment/__init__.py:
```diff
@@ -218,11 +220,14 @@
 __all__ = [
     "TabularMSA",
     "StripedSmithWaterman",
     "AlignmentStructure",
```
Just realized that there is an `AlignmentStructure` class. Outdated. Should be removed when we remove the SSW wrapper code. The new `AlignPath` class can replace it.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,305 @@
+        # TODO: Pending benchmark and optimization.
+
+        # Ensure bits is a 2D numpy array of ones and zeros.
+        if not isinstance(bits, np.ndarray) or bits.ndim != 2 or bits.shape[1] == 0:
```
`bits` may be array-like (e.g., a NumPy array or a Python list), and one can use `np.atleast_2d` to convert it into a 2-D array.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,305 @@
+        # Ensure bits is a 2D numpy array of ones and zeros.
+        if not isinstance(bits, np.ndarray) or bits.ndim != 2 or bits.shape[1] == 0:
+            raise ValueError("Input 'bits' must be a non-empty 2D numpy array.")
+        if not np.all(np.logical_or(bits == 0, bits == 1)):
```
This seems to be a brilliant solution. It outperforms the best GPT-4o could do (it suggests `np.isin(arr, [0, 1]).all()`).

Minor suggestion: `np.logical_or(bits == 0, bits == 1).all()` should be slightly faster than `np.all(...)`.

mataton replied:
Thank you!

mataton commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,305 @@
+        self.states = np.atleast_2d(np.asarray(states, dtype=np.uint8))
+
+        # start positions
+        self.starts = np.asarray(starts, dtype=int)
```
Is this okay as a regular `int`, or should we change this to `np.uint64` as well? As mentioned previously ~2Gbp is shorter than many genomes, and it is possible that the start position of a sequence will be beyond that point in the genome.

qiyunzhu replied:
Good point. I would suggest changing into `np.uint64` as well. Meanwhile, I don't think supporting negative start positions is necessary (I can't think of a use case).

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,341 @@
+        """
+        # Number of sequences needs to be explicitly provided, because the packed bits
+        # does not contain this information. (It is merely in multiples of 8.)
+        self.lengths = np.asarray(lengths, dtype=np.int64)
```
Good to know. Thanks for exploring!

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,410 @@
+            # Also need to ensure that the sequences provided actually match the
+            # lengths and states of the instantiated object.
+            byte_arr = np.stack([x._bytes for x in seqs])
+            gap_chars = [ord(x) for x in seqs[0].gap_chars]
```
I think that there is a pre-calculated class property `_gap_codes`.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,410 @@
+        gaps = np.squeeze(self.states)
+        codes = ["M", "I", "D", "P"]
+        if seqs is not None:
+            query = str(seqs[0])
```
Instead of comparing strings, we can directly compare the bytes underlying the strings.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,410 @@
+        # If `seqs` is provided, ensure that the sequence lengths match the alignment
+        # length.
+        if seqs:
+            if len(seqs[0]) != self.shape[1] or len(seqs[1]) != self.shape[1]:
```
I am under impression that this code assumes that the two sequences are of the same length. But in fact they don't have to be. The logic of this method may be revised.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,410 @@
+        if seqs is not None:
+            query = str(seqs[0])
+            ref = str(seqs[1])
+            for qchar, rchar in zip(query, ref):
```
See above. This code assumes that the two sequences are of the same length, which may not be the case.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,410 @@
+        lengths, gaps = cls._fix_arrays(lengths=np.array(lengths), gaps=np.array(gaps))
+        return cls(lengths, gaps, [0, 0])
+
+    def _run_length_encode(self, string_in):
```
This method does not use `self`. Therefore, it does not have to be a method of the class. Instead, it can be a standalone function.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,410 @@
+
+        return encoded_str
+
+    def _fix_arrays(lengths, gaps):
```
Likewise, this can be a standalone function.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,407 @@
+        seqs : list of skbio.sequence
+            Pair of aligned sequences to generate cigar string. If provided, will
+            distinguish match (=) and mismatch (X). Otherwise, will uniformly note
+            them as (mis)match (M).
```
Should note that seq1 is "query", seq2 is "reference".

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -146,13 +146,20 @@ def to_indices(self, gap=-1):
-        gap : -1, "del", or "mask", optional
-            If -1, replace gaps with this value. If "del", delete columns that have any
-            gap. If "mask", mask gaps. Default is -1.
+        gap : int, float, "del", or "mask", optional
```
I think an index cannot be a float. It is possible to cast floats into integers, but it shouldn't be a normal use case.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,446 @@
+            else:
+                raise ValueError("`seqs` must be of type string or Sequence object.")
+
+            idx1, idx2 = [int(x) for x in self.starts]
```
I thought that `starts` is always an integer array, therefore `int` is not necessary?

mataton replied:
`starts` is defined initially as `self.starts = np.asarray(starts, dtype=np.uint64)`, but when I unpacked `starts` with `idx1, idx2 = self.starts`, `idx1` and `idx2` became `float`s when we add them to `length` in line 335. Using `int` here solves this problem.

For example:
````python
>>> starts = np.array([0, 1], dtype=np.uint64)
>>> lengths = np.array([3, 4], dtype=np.int64)
>>> a, b = starts
>>> c, d = lengths
>>> type(a)
<class 'numpy.uint64'>
>>> type(c)
<class 'numpy.int64'>
>>> a + c
3.0
````

The other solution is to use `np.int64` instead of `np.uint64` when we define `starts` at the beginning of the class. `np.uint64` has strange behavior as I have mentioned elsewhere in this PR.

qiyunzhu replied:
Thanks for checking this! I think that using `np.int64` instead of `np.uint64` would be the best solution.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,446 @@
+                seq1 = seqs[0]._bytes
+                seq2 = seqs[1]._bytes
+            else:
+                raise ValueError("`seqs` must be of type string or Sequence object.")
```
Can be "`seqs` must be strings or Sequence objects."

mataton commented:
@qiyunzhu @mortonjt I believe this PR is ready for the final round of review. As an update from the initial round, there have been a few major changes. Most notably, all of the scoring functions such as `pairwise_align_scores` have been removed from this PR and will be addressed in a future separate PR. Additionally, the `to_cigar` method has been entirely rewritten. Other than those, the changes have addressed all comments from the previous rounds of review. Thank you for your help on this!

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,447 @@
+                    cigar.append(str(length) + "D")
+                    idx1 += length
+                elif state == 3:
+                    cigar.append(str(length) + "P")
```
There is a shorter may that can replace the code between lines 332 to 354:

```python
if state == 0:
    match_arr = ...
else:
    cigar.append(str(length) + codes[state])
flipped = state ^ 3
idx1 += flipped & 1 and length
idx2 += flipped & 2 and length
```

This method uses indexing and bitwise operations to avoid rewriting `cigar.append` and `idx +=` for each condition. It is not necessarily faster than the current solution, nor is it more elegant (bitwise operation is hard to understand!). Therefore, it is only for reference.

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -0,0 +1,447 @@
+        lengths = []
+        gaps = []
+        current_length = 0
+        mapping = {
```
Likewise, this `mapping` can be a constant outside the method or the class.

mortonjt reviewed (approved):
My comments have been addressed

qiyunzhu commented on skbio/alignment/_path.py:
```diff
@@ -237,29 +265,54 @@ def from_indices(cls, indices, gap=-1):
-            return cls.from_bits(indices == gap)
+            indices = np.asarray(indices)
+            return cls.from_bits(
+                indices == gap, [x[np.argmax(x != gap)] for x in indices]
```
@mataton After some exploration, the best non-Cython solution I learned is:

Replace:

```python
[x[np.argmax(x != gap)] for x in indices]
```

With:

```python
indices[np.arange(indices.shape[0]), np.argmax(indices != gap, axis=1)]
```

This [thread](https://stackoverflow.com/questions/16243955/numpy-first-occurrence-of-value-greater-than-existing-value) has some nice discussions and benchmarks. The most interesting thing is Mr Fooz's reply.
